core/src/main/scala/kafka/raft/KafkaFuturePurgatory.scala

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.raft
+
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.atomic.AtomicBoolean
+
+import kafka.server.{DelayedOperation, DelayedOperationPurgatory}
+import kafka.utils.Logging
+import kafka.utils.timer.Timer
+import org.apache.kafka.common.errors.TimeoutException
+import org.apache.kafka.raft.FuturePurgatory
+
+/**
+ * Simple purgatory shim for integration with the Raft library. We assume that
+ * both [[await()]] and [[completeAll()]] are called in the same thread.
+ */
+class KafkaFuturePurgatory[T](brokerId: Int, timer: Timer, reaperEnabled: Boolean = true)
+  extends FuturePurgatory[T] with Logging {
+
+  private val key = new Object()
+  private val purgatory = new DelayedOperationPurgatory[DelayedRaftRequest[T]](
+    "raft-request-purgatory", timer, brokerId, reaperEnabled = reaperEnabled)
+
+  override def await(future: CompletableFuture[T], maxWaitTimeMs: Long): Unit = {
+    val op = new DelayedRaftRequest(future, maxWaitTimeMs)
+    purgatory.tryCompleteElseWatch(op, Seq(key))
+    op.isCompletable.set(true)
+  }
+
+  override def completeAll(value: T): Unit = {
+    purgatory.checkAndComplete(key)
+  }
+
+  override def numWaiting(): Int = {
+    purgatory.numDelayed
+  }
+}
+
+class DelayedRaftRequest[T](future: CompletableFuture[T], delayMs: Long)
+  extends DelayedOperation(delayMs) {
+
+  val isCompletable = new AtomicBoolean(false)
+  val isExpired = new AtomicBoolean(false)
+
+  override def onExpiration(): Unit = {}
+
+  override def onComplete(): Unit = {
+    // the future may be completed by the caller thread already, in which case we can just skip here
+    if (future.isDone)
+      return
+
+    if (isExpired.get())
+      future.completeExceptionally(new TimeoutException("Request timed out in purgatory"))
+    else
+      future.complete(_)
```

Comment:
Hmm it doesn't seem like we have a way to pass through a value for the generic type `T`. I think that's why I did not make `KafkaFuturePurgatory` generic.

Reply:
Yeah when trying your suggesting to consolidate the purgatory and the treemap I also ran into this. After several failed tries I've come up with an approach. I'm honestly not sure if you all will like it though :P